R-package/R/callbacks.R

```diff
@@ -1,563 +1,923 @@
-#' Callback closures for booster training.
+.reserved_cb_names <- c("names", "class", "call", "params", "niter", "nfeatures", "folds")
+
+#' @title XGBoost Callback Constructor
+#' @description Constructor for defining the structure of callback functions that can be executed
+#' at different stages of model training (before / after training, before / after each boosting
+#' iteration).
+#' @param cb_name Name for the callback.
 #'
-#' These are used to perform various service tasks either during boosting iterations or at the end.
-#' This approach helps to modularize many of such tasks without bloating the main training methods,
-#' and it offers .
+#' If the callback produces some non-NULL result (from executing the function passed under
+#' `f_after_training`), that result will be added as an R attribute to the resulting booster
+#' (or as a named element in the result of CV), with the attribute name specified here.
 #'
-#' @details
-#' By default, a callback function is run after each boosting iteration.
-#' An R-attribute \code{is_pre_iteration} could be set for a callback to define a pre-iteration function.
+#' Names of callbacks must be unique - i.e. there cannot be two callbacks with the same name.
+#' @param env An environment object that will be passed to the different functions in the callback.
+#' Note that this environment will not be shared with other callbacks.
+#' @param f_before_training A function that will be executed before the training has started.
 #'
-#' When a callback function has \code{finalize} parameter, its finalizer part will also be run after
-#' the boosting is completed.
+#' If passing `NULL` for this or for the other function inputs, then no function will be executed.
 #'
-#' WARNING: side-effects!!! Be aware that these callback functions access and modify things in
-#' the environment from which they are called from, which is a fairly uncommon thing to do in R.
+#' If passing a function, it will be called with parameters supplied as non-named arguments
+#' matching the function signatures that are shown in the default value for each function argument.
+#' @param f_before_iter A function that will be executed before each boosting round.
```

Comment:
Does it return a boolean value?

Reply:
No, this function `f_before_iter` doesn't need to return anything if the user defines it.

The idea is that there can be cases in which some user function might need to execute before instead of after, or there might be actions that happen in both.

For example:
* A user callback that modifies things like the step sizes would likely fit better under `f_before_iter`.
* A user callback that uses some EM-type algorithm would likely fit better under `f_after_iter`.

Reply:
I'm mostly comparing the callback between R and Python, with the hope that we can minimize the difference when approapriate, the one in Python does return a boolean value for the same reason as the `f_after_iter`.

Reply:
Modified so that it would interpret the return value as a signal of whether to stop the training or not.